src/task.rs

```diff
@@ -319,6 +319,240 @@ impl<'a> Drop for CriticalSectionGuard<'a> {
+                }
+            }
+        }
+
+        impl Default for Config {
+            fn default() -> Self {
+                Self::default()
+            }
+        }
+
+        #[cfg(not(esp_idf_version_major = "4"))]
+        impl From<Config> for esp_task_wdt_config_t {
+            fn from(config: Config) -> Self {
+                esp_task_wdt_config_t {
+                    timeout_ms: config.duration.as_millis() as u32,
+                    trigger_panic: config.panic_on_trigger,
+                    idle_core_mask: config
+                        .subscribed_idle_tasks
+                        .iter()
+                        .fold(0u32, |mask, core| mask & (1 << (*core as u32))),
+                }
+            }
+        }
+    }
+
+    pub trait Twdt {}
+
+    pub struct TWDTDriver<'d> {
+        init_by_idf: bool,
+        _marker: PhantomData<&'d mut ()>,
+    }
+
+    static TWDT_DRIVER_REF_COUNT: AtomicUsize = AtomicUsize::new(0);
+
+    impl TWDTDriver<'_> {
+        fn init(config: &config::Config) -> Result<Self, EspError> {
+            TWDT_DRIVER_REF_COUNT.fetch_add(1, Ordering::SeqCst);
+            let init_by_idf = Self::watchdog_is_init_by_idf();
+
+            #[cfg(not(esp_idf_version_major = "4"))]
+            if !init_by_idf {
+                unsafe {
+                    esp!(esp_task_wdt_init(config.into()))?;
+                }
+            } else {
+                unsafe {
+                    esp!(esp_task_wdt_reconfigure(config.into()))?;
+                }
+            }
+
+            #[cfg(esp_idf_version_major = "4")]
+            unsafe {
+                esp!(esp_task_wdt_init(
+                    config.duration.as_secs() as u32,
+                    config.panic_on_trigger
+                ))?;
+            }
+
+            Ok(Self {
+                init_by_idf,
+                _marker: Default::default(),
+            })
+        }
+    }
+
+    impl<'d> TWDTDriver<'d> {
+        pub fn new(
+            _twdt: impl Peripheral<P = TWDT> + 'd,
+            config: &config::Config,
+        ) -> Result<Self, EspError> {
+            Self::init(config)
+        }
+
+        pub fn watch_current_task(&mut self) -> Result<WatchdogSubscription<'_>, EspError> {
+            unsafe { esp!(esp_task_wdt_add(core::ptr::null_mut())) }?;
+            Ok(WatchdogSubscription::new())
+        }
+
+        fn watchdog_is_init_by_idf() -> bool {
+            if cfg!(not(any(
+                esp_idf_version_major = "4",
+                esp_idf_version = "5.0"
+            ))) {
+                cfg!(esp_idf_esp_task_wdt_init)
+            } else {
+                #[cfg(any(esp_idf_version_major = "4", esp_idf_version = "5.0"))]
+                !matches!(
+                    unsafe { esp_task_wdt_status(core::ptr::null_mut()) },
+                    ESP_ERR_INVALID_STATE
+                )
+            }
+        }
+
+        fn deinit(&self) {
+            if !self.init_by_idf {
```

Comment:
OK so what you are implementing is sort of an "Option 3" which is - keep the TWDT running, if it was running before.

This is easy to implement, but kind of hard to explain to users w.r.t. watching idle tasks. As in, if the watchdog was **not** running before `TWDTDriver`, when `TWDTDriver` is destroyed, idle tasks will no longer be watched, even if the user had explicitly requested so. And the other way around: if the watchdog was running, idle tasks will continue to be watched even when the `TWDTDriver` is destroyed.

So I have a suspicion that we have to change this in future. With that said, I'm fine with keeping this behavior for now, in the name of simplicity.

Reply:
I think we can apply your Option 1. This will simplify the behavior as you said and will work with no panics.

Reply:
Let's stick with your Option 3 for now. I'm drifting away from Option 1 in favor of Option 2. Not watching idle tasks and making these starve for a long time (and that would happen with my Option 1 if user is not careful to keep a `TWDTDriver` instance forever) might have bad consequences. To quote ESP IDF:

```
NOTE: The idle task is responsible for freeing the kernel allocated memory from tasks that have been deleted. It is therefore important that the idle task is not starved of microcontroller processing time if your application makes any calls to vTaskDelete (). Memory allocated by the task code is not automatically freed, and should be freed before the task is deleted.
```

The above also means by the way, **that doing computationally-intensive stuff FOREVER in a loop and calling `WatchdogSubscription::feed` from time to time might not be such a great idea after all**. You'll need to yield in the loop as well, for at least 10ms!